prow/pod-utils/gcs/upload_test.go

```diff
@@ -30,78 +29,138 @@ import (
+			},
 		},
 		{
-			name:           "all failing",
-			passingTargets: 0,
-			failingTargets: 10,
-			expectedErr:    true,
+			name: "some failed",
+			destUploadBehaviors: []destUploadBehavior{
+				{
+					dest:     "some-failed-dest1",
+					doesPass: true,
+					isFlaky:  false,
+				},
+				{
+					dest:     "some-failed-dest2",
+					doesPass: false,
+					isFlaky:  false,
+				},
+			},
 		},
 	}
 
 	for _, testCase := range testCases {
-		lock := sync.Mutex{}
-		count := 0
 
-		update := func() {
-			lock.Lock()
-			defer lock.Unlock()
-			count = count + 1
-		}
+		uploadFuncs := map[string]UploadFunc{}
 
-		fail := func(_ dataWriter) error {
-			update()
-			return errors.New("fail")
-		}
+		currentTestStates := map[string]destUploadBehavior{}
+		currentTestStatesLock := sync.Mutex{}
 
-		success := func(_ dataWriter) error {
-			update()
-			return nil
-		}
+		for _, destBehavior := range testCase.destUploadBehaviors {
 
-		targets := map[string]UploadFunc{}
-		for i := 0; i < testCase.passingTargets; i++ {
-			targets[fmt.Sprintf("pass-%d", i)] = success
-		}
+			currentTestStates[destBehavior.dest] = destBehavior
 
-		for i := 0; i < testCase.failingTargets; i++ {
-			targets[fmt.Sprintf("fail-%d", i)] = fail
-		}
+			getUploadFunc := func(destBehavior destUploadBehavior) UploadFunc {
+
+				return func(writer dataWriter) error {
+					currentTestStatesLock.Lock()
+					defer currentTestStatesLock.Unlock()
+
+					currentDestUploadBehavior := currentTestStates[destBehavior.dest]
+
+					if !currentDestUploadBehavior.doesPass {
+						return fmt.Errorf("%v: %v failed", testCase.name, destBehavior.dest)
+					}
+
+					if currentDestUploadBehavior.isFlaky {
+						currentDestUploadBehavior.isFlaky = false
+						currentTestStates[destBehavior.dest] = currentDestUploadBehavior
+						return fmt.Errorf("%v: %v flaky", testCase.name, destBehavior.dest)
+					}
+
+					delete(currentTestStates, destBehavior.dest)
+					return nil
+				}
+			}
+
+			uploadFuncs[destBehavior.dest] = getUploadFunc(destBehavior)
 
-		err := Upload("", "", "", targets)
-		if err != nil && !testCase.expectedErr {
-			t.Errorf("%s: expected no error but got %v", testCase.name, err)
 		}
-		if err == nil && testCase.expectedErr {
-			t.Errorf("%s: expected an error but got none", testCase.name)
+
+		err := Upload("", "", "", uploadFuncs)
+
+		isErrExpected := false
+		for _, currentTestState := range currentTestStates {
+
+			if currentTestState.doesPass {
+				t.Errorf("%v: %v did not get uploaded", testCase.name, currentTestState.dest)
+				break
+			}
+
+			if !isErrExpected && !currentTestState.doesPass {
+				isErrExpected = true
+			}
 		}
 
-		if count != (testCase.passingTargets + testCase.failingTargets) {
-			t.Errorf("%s: had %d passing and %d failing targets but only ran %d targets, not %d", testCase.name, testCase.passingTargets, testCase.failingTargets, count, testCase.passingTargets+testCase.failingTargets)
+		if err != nil && !isErrExpected {
```

Comment:
Sorry didn't see this on last review, is there a test case missing that covers `err == nil && isErrExpected`?